pystac/extensions/base.py

```diff
@@ -132,13 +135,47 @@ def remove_from(cls, obj: S) -> None:
             ]
 
     @classmethod
-    def has_extension(cls, obj: S) -> bool:
+    def has_extension(cls, obj: S, version_prefix: Optional[str] = None) -> bool:
         """Check if the given object implements this extension by checking
-        :attr:`pystac.STACObject.stac_extensions` for this extension's schema URI."""
-        return (
-            obj.stac_extensions is not None
-            and cls.get_schema_uri() in obj.stac_extensions
-        )
+        :attr:`pystac.STACObject.stac_extensions` for this extension's schema
+        URI. Optionally check for a minimum or specific semver of the extension
+        with ``version_prefix``."""
+        version = cls.get_version(obj)
+        if version and version_prefix:
+            return bool(version) and version.startswith(version_prefix)
+        else:
+            return bool(version)
+
+    @classmethod
+    def get_version(cls, obj: S) -> Optional[str]:
+        """Gets the version of this extension from the semver portion of the
+        extension schema URI in :attr:`pystac.STACObject.stac_extensions` for
+        the given object."""
+        if obj.stac_extensions is not None:
+            parsed_uri = safe_urlparse(cls.get_schema_uri())
+            uri_path_parts = Path(parsed_uri.path).parts
+            uri_path_prefix = join_path_or_url(
+                JoinType.from_parsed_uri(parsed_uri), *uri_path_parts[:-2]
+            )
+            uri_prefix = urlunparse(
+                (
+                    parsed_uri.scheme,
+                    parsed_uri.netloc,
+                    uri_path_prefix,
+                    None,
+                    None,
+                    None,
+                )
+            )
+            uri = next(
+                (uri for uri in obj.stac_extensions if uri.startswith(uri_prefix)), None
+            )
+            if uri is None:
+                return None
+            else:
+                return Path(safe_urlparse(uri).path).parts[-2]
+        else:
+            return None
```

Comment:
I think these are good changes, but w/ my suggested change to https://github.com/stac-utils/pystac/pull/1006/files#diff-28af67c49c05ff66bbfd9f3ab4487687c657a3c09d640b93ec9401b76ffb9665R687 we won't need them. Per your comment https://github.com/stac-utils/pystac/pull/1006#issuecomment-1452804629, maybe these should be part of a larger reimagining of the extension detection architecture?

Reply:
I've removed this. Agreed, if nothing else, a good exercise in thinking about what would be required to handle different extension versions.